annif/backend/yake.py

```diff
@@ -0,0 +1,196 @@
+    def default_params(self):
+        params = backend.AnnifBackend.DEFAULT_PARAMETERS.copy()
+        params.update(self.DEFAULT_PARAMETERS)
+        return params
+
+    @property
+    def is_trained(self):
+        return True
+
+    @property
+    def label_types(self):
+        mapping = {'pref': SKOS.prefLabel,
+                   'alt': SKOS.altLabel,
+                   'hidden': SKOS.hiddenLabel}
+        if 'label_types' in self.params:
+            lt_entries = self.params['label_types'].split(',')
+            try:
+                return [mapping[lt.strip()] for lt in lt_entries]
+            except KeyError as err:
+                raise ConfigurationException(
+                    f'invalid label type {err}', backend_id=self.backend_id)
+        else:
+            return [mapping[lt] for lt in self.params['default_label_types']]
+
+    @property
+    def graph(self):
+        if self._graph is None:
+            self.info('Loading graph')
+            self._graph = self.project.vocab.as_graph()
+        return self._graph
+
+    def initialize(self):
+        self._initialize_index()
+        self._kw_extractor = yake.KeywordExtractor(
+            lan=self.params['language'],
+            n=self.params['max_ngram_size'],
+            dedupLim=self.params['deduplication_threshold'],
+            dedupFunc=self.params['deduplication_algo'],
+            windowsSize=self.params['window_size'],
+            top=self.params['num_keywords'],
+            features=self.params['features'])
+
+    def _initialize_index(self):
+        if self._index is None:
+            path = os.path.join(self.datadir, self.INDEX_FILE)
+            if os.path.exists(path):
+                self._index = self._load_index(path)
+                self.info(
+                    f'Loaded index from {path} with {len(self._index)} labels')
+            else:
+                self.info('Creating index')
+                self._create_index()
+                self._save_index(path)
+                self.info(f'Created index with {len(self._index)} labels')
+
+    def _save_index(self, path):
+        with open(path, 'w', encoding='utf-8') as indexfile:
+            for label, uris in self._index.items():
+                line = label + '\t' + ' '.join(uris)
+                print(line, file=indexfile)
+
+    def _load_index(self, path):
+        index = dict()
+        with open(path, 'r', encoding='utf-8') as indexfile:
+            for line in indexfile:
+                label, uris = line.strip().split('\t')
+                index[label] = uris.split()
+        return index
+
+    def _create_index(self):
+        index = defaultdict(set)
+        for concept in self.graph.subjects(RDF.type, SKOS.Concept):
+            if (concept, OWL.deprecated, rdflib.Literal(True)) in self.graph:
+                continue
+            uri = str(concept)
+            for label_type in self.label_types:
+                for label in self.graph.objects(concept, label_type):
```

Comment:
Using `graph.preferredLabel` could perhaps be a bit more elegant here, it can be given the list of properties (label types) and the language. See e.g. [this example](https://github.com/NatLibFi/Annif/pull/462/files#diff-231863667a713a02bb9fc65f1bda5432e5a897cc1c97260ed381443be039a82cR181) in the MLLM backend

Reply:
I made a helper method `_get_concept_labels` not using the `graph.preferredLabel` because that always returns only one label (the first match that is encountered when comparing against the given label types in `labelProperties`). And I think here it's better to have all the requested label types, so it seems that looping over label types would be needed even with `graph.preferredLabel`.

Reply:
Ah, you're right, `graph.preferredLabel` wouldn't work in this case.